## Data paths

In [1]:
DATA_PATH = r"D:\datasets_complete\datasets-projects\online+retail+ii\online_retail_II.xlsx"

# Data Variable Information
**InvoiceNo**: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.

**StockCode**: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.

**Description**: Product (item) name. Nominal.

**Quantity**: The quantities of each product (item) per transaction. Numeric.

**InvoiceDate**: Invice date and time. Numeric. The day and time when a transaction was generated.

**UnitPrice**: Unit price. Numeric. Product price per unit in sterling (£).
 
*CustomerID*: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.

**Country**: Country name. Nominal. The name of the country where a customer resides.

[Link to info](https://archive.ics.uci.edu/dataset/502/online+retail+ii)

# Imports

In [2]:
import pandas as pd
import numpy as np

## Data read and dataframe creation (`PANDAS`)

In [ ]:
# df = pd.read_excel(DATA_PATH, sheet_name="Year 2010-2011")

In [ ]:
# Saved to avoid re-loading the excel file.
# df.to_pickle('df_2010-2011.pkl')

In [47]:
# Load the pre-saved pickle file
df = pd.read_pickle('df_2010-2011.pkl')
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [49]:
display(df.describe())
df.describe(include=['O'])

Quantity                    InvoiceDate          Price  \
count  541910.000000                         541910  541910.000000   
mean        9.552234  2011-07-04 13:35:22.342307584       4.611138   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.080957                            NaN      96.759765   

         Customer ID  
count  406830.000000  
mean    15287.684160  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.603074

Invoice StockCode                         Description         Country
count    541910    541910                              540456          541910
unique    25900      4070                                4223              38
top      573585    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq       1114      2313                                2369          495478

In [50]:
# Normalization of columns names
df.columns = [col.lower().replace(' ', '_') for col in df.columns]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   invoice      541910 non-null  object        
 1   stockcode    541910 non-null  object        
 2   description  540456 non-null  object        
 3   quantity     541910 non-null  int64         
 4   invoicedate  541910 non-null  datetime64[ns]
 5   price        541910 non-null  float64       
 6   customer_id  406830 non-null  float64       
 7   country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [51]:
df_column_names = df.columns.to_list()
df_column_names

['invoice',
 'stockcode',
 'description',
 'quantity',
 'invoicedate',
 'price',
 'customer_id',
 'country']

# EDA Notes

Client Clustering

- `541910` entries, `Description` and `Customer ID` do not match the number of entries (missing values).
- Negative values in `Quantity`: `-80995.000000`. It matches the **max** amount.
- Negative values in `Price`: `-11062.060000`.
- `25900` unique `Invoice` values.
- `4070` unique `StockCode` values, meaning 4070 products; but `4223` `Description` values. 

### `invoice` column

In [55]:
df[df['quantity']<0].head() # Overview of `quantity` negative values

invoice stockcode                       description  quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            invoicedate  price  customer_id         country  
141 2010-12-01 09:41:00  27.50      14527.0  United Kingdom  
154 2010-12-01 09:49:00   4.65      15311.0  United Kingdom  
235 2010-12-01 10:24:00   1.65      17548.0  United Kingdom  
236 2010-12-01 10:24:00   0.29      17548.0  United Kingdom  
237 2010-12-01 10:24:00   0.29      17548.0  United Kingdom

In [72]:
df['invoice'] = df['invoice'].astype('str')
df_cancellation_invoices = df[df['invoice'].str.startswith('C')]
# df[df['invoice'].str.match("^\\d{6}$") == False] # Regex for NOT 6 consecutive numbers as invoice id
display(df_cancellation_invoices.head())
print("Invoices that are cancellations:", df_cancellation_invoices.shape[0]) # According to the data information website

invoice stockcode                       description  quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            invoicedate  price  customer_id         country  
141 2010-12-01 09:41:00  27.50      14527.0  United Kingdom  
154 2010-12-01 09:49:00   4.65      15311.0  United Kingdom  
235 2010-12-01 10:24:00   1.65      17548.0  United Kingdom  
236 2010-12-01 10:24:00   0.29      17548.0  United Kingdom  
237 2010-12-01 10:24:00   0.29      17548.0  United Kingdom

Invoices that are cancellations: 9288


In [75]:
# Other letters in 'invoice' column
df['invoice'].str.replace("[0-9]", "", regex=True).unique()
# Output: There are invoices that start with 'C' and 'A' letters.

array(['', 'C', 'A'], dtype=object)

In [87]:
display(df[df['invoice'].str.startswith('A')].head()) # Overview of `A` invoices
# These are adjustments invoices. Erase these rows.
# List indexes to drop
indexes_to_drop = df[df['invoice'].str.startswith('A')].index
# display(indexes_to_drop)

invoice stockcode      description  quantity         invoicedate  \
299982  A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   
299983  A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           price  customer_id         country  
299982  11062.06          NaN  United Kingdom  
299983 -11062.06          NaN  United Kingdom  
299984 -11062.06          NaN  United Kingdom

## `stockcode` column

In [83]:
df['stockcode'] = df['stockcode'].astype('str')

In [91]:
stockcodes_abnormal = df[(df['stockcode'].str.match("^\\d{5}$") == False) & (df['stockcode'].str.match("^\\d{5}[a-zA-Z]+$") == False)]['stockcode'].unique()
# Regex for NOT 5 consecutive numbers or 5 consecutive numbers followed by letters in the values of 'stockcode' column
print("Abnormal stockcodes:", stockcodes_abnormal)
# Output: There are some abnormal stockcodes

Abnormal stockcodes: ['POST' 'D' 'C2' 'DOT' 'M' 'BANK CHARGES' 'S' 'AMAZONFEE' 'DCGS0076'
 'DCGS0003' 'gift_0001_40' 'DCGS0070' 'm' 'gift_0001_50' 'gift_0001_30'
 'gift_0001_20' 'DCGS0055' 'DCGS0072' 'DCGS0074' 'DCGS0069' 'DCGS0057'
 'DCGSSBOY' 'DCGSSGIRL' 'gift_0001_10' 'PADS' 'DCGS0004' 'DCGS0073'
 'DCGS0071' 'DCGS0066P' 'DCGS0068' 'DCGS0067' 'B' 'CRUK']


In [94]:
# Show entries witch each abnormal stockcode in stockcodes_abnormal
for code in stockcodes_abnormal:
    display(df[df['stockcode'] == code].head(10))

invoice stockcode description  quantity         invoicedate  price  \
45    536370      POST     POSTAGE         3 2010-12-01 08:45:00   18.0   
386   536403      POST     POSTAGE         1 2010-12-01 11:27:00   15.0   
1123  536527      POST     POSTAGE         1 2010-12-01 13:04:00   18.0   
5095  536840      POST     POSTAGE         1 2010-12-02 18:27:00   18.0   
5258  536852      POST     POSTAGE         1 2010-12-03 09:51:00   18.0   
5325  536858      POST     POSTAGE         2 2010-12-03 10:36:00   40.0   
5369  536861      POST     POSTAGE         3 2010-12-03 10:44:00   18.0   
6602  536967      POST     POSTAGE         1 2010-12-03 12:57:00   18.0   
6676  536974      POST     POSTAGE         2 2010-12-03 13:59:00   18.0   
6973  536983      POST     POSTAGE         1 2010-12-03 14:30:00   18.0   

      customer_id      country  
45        12583.0       France  
386       12791.0  Netherlands  
1123      12662.0      Germany  
5095      12738.0      Germany  
5258      12686.0       France  
5325      13520.0  Switzerland  
5369      12427.0      Germany  
6602      12600.0      Germany  
6676      12682.0       France  
6973      12712.0      Germany

invoice stockcode description  quantity         invoicedate   price  \
141    C536379         D    Discount        -1 2010-12-01 09:41:00   27.50   
9038   C537164         D    Discount        -1 2010-12-05 13:21:00   29.29   
14498  C537597         D    Discount        -1 2010-12-07 12:34:00  281.00   
19392  C537857         D    Discount        -1 2010-12-08 16:00:00  267.12   
31134  C538897         D    Discount        -1 2010-12-15 09:14:00    5.76   
31135  C538897         D    Discount        -1 2010-12-15 09:14:00   42.50   
31663  C539003         D    Discount        -1 2010-12-15 12:35:00   26.93   
38609  C539589         D    Discount        -1 2010-12-20 13:14:00   13.88   
44405  C540171         D    Discount        -1 2011-01-05 12:26:00   22.97   
88032  C543752         D    Discount        -1 2011-02-11 14:12:00   64.27   

       customer_id         country  
141        14527.0  United Kingdom  
9038       14527.0  United Kingdom  
14498      15498.0  United Kingdom  
19392      17340.0  United Kingdom  
31134      16422.0  United Kingdom  
31135      16422.0  United Kingdom  
31663      14527.0  United Kingdom  
38609      14527.0  United Kingdom  
44405      14527.0  United Kingdom  
88032      14031.0  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
1423   536540        C2    CARRIAGE         1 2010-12-01 14:05:00   50.0   
12119  537368        C2    CARRIAGE         1 2010-12-06 12:40:00   50.0   
12452  537378        C2    CARRIAGE         1 2010-12-06 13:06:00   50.0   
19975  537963        C2    CARRIAGE         1 2010-12-09 11:30:00   50.0   
20016  538002        C2    CARRIAGE         1 2010-12-09 11:48:00   50.0   
34369  539337        C2    CARRIAGE         1 2010-12-17 10:46:00   50.0   
34725  539421        C2    CARRIAGE         1 2010-12-17 14:21:00   50.0   
37644  539473        C2    CARRIAGE         1 2010-12-19 14:24:00   50.0   
39777  539688        C2    CARRIAGE         1 2010-12-21 11:00:00  150.0   
42332  539984        C2    CARRIAGE         1 2010-12-23 14:58:00   50.0   

       customer_id          country  
1423       14911.0             EIRE  
12119      14911.0             EIRE  
12452      14911.0             EIRE  
19975      13369.0   United Kingdom  
20016      14932.0  Channel Islands  
34369          NaN             EIRE  
34725      14016.0             EIRE  
37644      14911.0             EIRE  
39777      12678.0           France  
42332      14911.0             EIRE

invoice stockcode     description  quantity         invoicedate   price  \
1815   536544       DOT  DOTCOM POSTAGE         1 2010-12-01 14:32:00  569.77   
3041   536592       DOT  DOTCOM POSTAGE         1 2010-12-01 17:06:00  607.49   
5450   536862       DOT  DOTCOM POSTAGE         1 2010-12-03 11:13:00  254.43   
5545   536864       DOT  DOTCOM POSTAGE         1 2010-12-03 11:27:00  121.06   
5686   536865       DOT  DOTCOM POSTAGE         1 2010-12-03 11:28:00  498.47   
6165   536876       DOT  DOTCOM POSTAGE         1 2010-12-03 11:36:00  887.52   
10812  537237       DOT  DOTCOM POSTAGE         1 2010-12-06 09:58:00  863.74   
11381  537240       DOT  DOTCOM POSTAGE         1 2010-12-06 10:08:00  940.87   
13924  537434       DOT  DOTCOM POSTAGE         1 2010-12-06 16:57:00  950.99   
15650  537638       DOT  DOTCOM POSTAGE         1 2010-12-07 15:28:00  836.14   

       customer_id         country  
1815           NaN  United Kingdom  
3041           NaN  United Kingdom  
5450           NaN  United Kingdom  
5545           NaN  United Kingdom  
5686           NaN  United Kingdom  
6165           NaN  United Kingdom  
10812          NaN  United Kingdom  
11381          NaN  United Kingdom  
13924          NaN  United Kingdom  
15650          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
2239  536569         M      Manual         1 2010-12-01 15:35:00   1.25   
2250  536569         M      Manual         1 2010-12-01 15:35:00  18.95   
5685  536865         M      Manual         1 2010-12-03 11:28:00   2.55   
6798  536981         M      Manual         2 2010-12-03 14:26:00   0.85   
7976  537077         M      Manual        12 2010-12-05 11:59:00   0.42   
8530  537137         M      Manual        36 2010-12-05 12:43:00   0.85   
8587  537140         M      Manual         1 2010-12-05 12:53:00   0.42   
8588  537140         M      Manual         1 2010-12-05 12:53:00   0.85   
9326  537199         M      Manual         1 2010-12-05 14:06:00   0.65   
9575  537208         M      Manual         4 2010-12-05 15:12:00   0.85   

      customer_id         country  
2239      16274.0  United Kingdom  
2250      16274.0  United Kingdom  
5685          NaN  United Kingdom  
6798      14723.0  United Kingdom  
7976      17062.0  United Kingdom  
8530      16327.0  United Kingdom  
8587      12748.0  United Kingdom  
8588      12748.0  United Kingdom  
9326      15894.0  United Kingdom  
9575      15889.0  United Kingdom

invoice     stockcode   description  quantity         invoicedate  \
4406     536779  BANK CHARGES  Bank Charges         1 2010-12-02 15:08:00   
14435   C537572  BANK CHARGES  Bank Charges        -1 2010-12-07 12:00:00   
28992   C538680  BANK CHARGES  Bank Charges        -1 2010-12-13 17:10:00   
62508    541505  BANK CHARGES  Bank Charges         1 2011-01-18 15:58:00   
64573   C541653  BANK CHARGES  Bank Charges        -1 2011-01-20 11:50:00   
90558   C544049  BANK CHARGES  Bank Charges        -1 2011-02-15 12:39:00   
96675   C544575  BANK CHARGES  Bank Charges        -1 2011-02-21 13:58:00   
96676   C544576  BANK CHARGES  Bank Charges        -1 2011-02-21 14:01:00   
96715   C544584  BANK CHARGES  Bank Charges        -1 2011-02-21 14:52:00   
112233  C545887  BANK CHARGES  Bank Charges        -1 2011-03-07 15:54:00   

          price  customer_id         country  
4406      15.00      15823.0  United Kingdom  
14435     95.38          NaN  United Kingdom  
28992    966.92          NaN  United Kingdom  
62508     15.00      15939.0  United Kingdom  
64573   1050.15          NaN  United Kingdom  
90558    566.37          NaN  United Kingdom  
96675    134.76          NaN  United Kingdom  
96676    149.16          NaN  United Kingdom  
96715    109.84          NaN  United Kingdom  
112233   326.68          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
14436  C537581         S     SAMPLES        -1 2010-12-07 12:03:00  12.95   
14437  C537581         S     SAMPLES        -1 2010-12-07 12:03:00  52.00   
96680  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   5.74   
96681  C544580         S     SAMPLES        -1 2011-02-21 14:25:00  11.08   
96682  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   5.79   
96683  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   6.70   
96684  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   8.74   
96685  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   9.74   
96686  C544580         S     SAMPLES        -1 2011-02-21 14:25:00   9.99   
96687  C544580         S     SAMPLES        -1 2011-02-21 14:25:00  20.98   

       customer_id         country  
14436          NaN  United Kingdom  
14437          NaN  United Kingdom  
96680          NaN  United Kingdom  
96681          NaN  United Kingdom  
96682          NaN  United Kingdom  
96683          NaN  United Kingdom  
96684          NaN  United Kingdom  
96685          NaN  United Kingdom  
96686          NaN  United Kingdom  
96687          NaN  United Kingdom

invoice  stockcode description  quantity         invoicedate     price  \
14514  C537600  AMAZONFEE  AMAZON FEE        -1 2010-12-07 12:41:00      1.00   
15016  C537630  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:04:00  13541.33   
15017   537632  AMAZONFEE  AMAZON FEE         1 2010-12-07 15:08:00  13541.33   
16002  C537644  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:34:00  13474.79   
16313  C537647  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:41:00   5519.25   
16356  C537651  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:49:00  13541.33   
16357  C537652  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:51:00   6706.71   
43702  C540117  AMAZONFEE  AMAZON FEE        -1 2011-01-05 09:55:00  16888.02   
43703  C540118  AMAZONFEE  AMAZON FEE        -1 2011-01-05 09:57:00  16453.71   
96844  C544587  AMAZONFEE  AMAZON FEE        -1 2011-02-21 15:07:00   5575.28   

       customer_id         country  
14514          NaN  United Kingdom  
15016          NaN  United Kingdom  
15017          NaN  United Kingdom  
16002          NaN  United Kingdom  
16313          NaN  United Kingdom  
16356          NaN  United Kingdom  
16357          NaN  United Kingdom  
43702          NaN  United Kingdom  
43703          NaN  United Kingdom  
96844          NaN  United Kingdom

invoice stockcode                   description  quantity  \
21326  538071  DCGS0076  SUNJAR LED NIGHT NIGHT LIGHT         1   
39313  539631  DCGS0076  SUNJAR LED NIGHT NIGHT LIGHT         2   

              invoicedate  price  customer_id         country  
21326 2010-12-09 14:09:00  16.13          NaN  United Kingdom  
39313 2010-12-20 15:03:00  16.13          NaN  United Kingdom

invoice stockcode          description  quantity         invoicedate  \
24906   538349  DCGS0003  BOXED GLASS ASHTRAY         1 2010-12-10 14:59:00   
36460   539451  DCGS0003  BOXED GLASS ASHTRAY         1 2010-12-17 16:59:00   
76251   542622  DCGS0003  BOXED GLASS ASHTRAY         1 2011-01-31 09:09:00   
176006  551995  DCGS0003  BOXED GLASS ASHTRAY         1 2011-05-05 15:42:00   
279258  561256  DCGS0003                 ebay        -7 2011-07-26 11:53:00   

        price  customer_id         country  
24906    2.51          NaN  United Kingdom  
36460    2.51          NaN  United Kingdom  
76251    2.46          NaN  United Kingdom  
176006   2.46          NaN  United Kingdom  
279258   0.00          NaN  United Kingdom

invoice     stockcode                         description  quantity  \
38248   539492  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   
145463  548893  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   
281008  561513  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   

               invoicedate  price  customer_id         country  
38248  2010-12-20 10:14:00  34.04          NaN  United Kingdom  
145463 2011-04-04 15:54:00  33.33          NaN  United Kingdom  
281008 2011-07-27 15:12:00  33.33          NaN  United Kingdom

invoice stockcode            description  quantity         invoicedate  \
40052   539718  DCGS0070  CAMOUFLAGE DOG COLLAR         1 2010-12-21 13:06:00   
279253  561251  DCGS0070                    NaN        -8 2011-07-26 11:51:00   

        price  customer_id         country  
40052   12.72          NaN  United Kingdom  
279253   0.00          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
40383  539736         m      Manual         1 2010-12-21 15:18:00   2.55   

       customer_id         country  
40383          NaN  United Kingdom

invoice     stockcode                         description  quantity  \
42057   539958  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
178559  552232  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
239744  558066  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
245516  558614  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   

               invoicedate  price  customer_id         country  
42057  2010-12-23 13:26:00  42.55          NaN  United Kingdom  
178559 2011-05-06 15:54:00  41.67          NaN  United Kingdom  
239744 2011-06-24 15:45:00  41.67          NaN  United Kingdom  
245516 2011-06-30 15:56:00  41.67          NaN  United Kingdom

invoice     stockcode                         description  quantity  \
44725   540238  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
94054   544323  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
95034   544434  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
208808  555149  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
297099  562933  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
317442  564760  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   
317501  564761  gift_0001_30                                 NaN        30   
322164  565231  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00         1   

               invoicedate  price  customer_id         country  
44725  2011-01-05 14:44:00  25.53          NaN  United Kingdom  
94054  2011-02-17 15:51:00  25.00          NaN  United Kingdom  
95034  2011-02-18 16:12:00  25.00          NaN  United Kingdom  
208808 2011-05-31 15:49:00  25.00          NaN  United Kingdom  
297099 2011-08-10 16:51:00  25.00          NaN  United Kingdom  
317442 2011-08-30 10:47:00  25.00          NaN  United Kingdom  
317501 2011-08-30 10:48:00   0.00          NaN  United Kingdom  
322164 2011-09-02 09:26:00  25.00          NaN  United Kingdom

invoice     stockcode                          description  quantity  \
44794   540238  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
55589   540995  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
91278   544089  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
161388  550474  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         2   
163440  550542  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
228814  556955  gift_0001_20  to push order througha s stock was         10   
234138  557500  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
239903  558068  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
291036  562420  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   
431349  573585  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00         1   

               invoicedate  price  customer_id         country  
44794  2011-01-05 14:44:00  17.02          NaN  United Kingdom  
55589  2011-01-13 09:30:00  16.67          NaN  United Kingdom  
91278  2011-02-15 17:51:00  16.67          NaN  United Kingdom  
161388 2011-04-18 13:58:00  16.67          NaN  United Kingdom  
163440 2011-04-19 11:37:00  16.67          NaN  United Kingdom  
228814 2011-06-16 09:04:00   0.00          NaN  United Kingdom  
234138 2011-06-20 15:27:00  16.67          NaN  United Kingdom  
239903 2011-06-24 15:51:00  16.67          NaN  United Kingdom  
291036 2011-08-04 16:38:00  16.67          NaN  United Kingdom  
431349 2011-10-31 14:41:00  16.67          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
74837  542529  DCGS0055         NaN        -1 2011-01-28 13:08:00    0.0   

       customer_id         country  
74837          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
74838  542531  DCGS0072         NaN        -1 2011-01-28 13:08:00    0.0   

       customer_id         country  
74838          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
74839  542532  DCGS0074         NaN        -1 2011-01-28 13:09:00    0.0   

       customer_id         country  
74839          NaN  United Kingdom

invoice stockcode            description  quantity         invoicedate  \
75053   542541  DCGS0069  OOH LA LA DOGS COLLAR         1 2011-01-28 14:25:00   
279255  561252  DCGS0069                   ebay        -5 2011-07-26 11:52:00   

        price  customer_id         country  
75053   15.79          NaN  United Kingdom  
279255   0.00          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
75295  542582  DCGS0057         NaN        -6 2011-01-28 15:01:00    0.0   

       customer_id         country  
75295          NaN  United Kingdom

invoice stockcode     description  quantity         invoicedate  price  \
84016   543358  DCGSSBOY  BOYS PARTY BAG         1 2011-02-07 14:04:00   3.29   
97246   544599  DCGSSBOY  BOYS PARTY BAG         1 2011-02-21 18:15:00   3.29   
112723  545897  DCGSSBOY  BOYS PARTY BAG         5 2011-03-07 17:22:00   3.29   
116891  546306  DCGSSBOY  BOYS PARTY BAG         1 2011-03-10 16:16:00   3.29   
128107  547249  DCGSSBOY  BOYS PARTY BAG         6 2011-03-22 09:27:00   3.29   
160487  550458  DCGSSBOY  BOYS PARTY BAG         1 2011-04-18 13:13:00   3.29   
178669  552234  DCGSSBOY  BOYS PARTY BAG         7 2011-05-06 16:01:00   3.29   
262771  559923  DCGSSBOY  BOYS PARTY BAG        20 2011-07-13 16:07:00   3.29   
278378  561209  DCGSSBOY  BOYS PARTY BAG         2 2011-07-25 16:57:00   1.25   
365966  568716  DCGSSBOY  BOYS PARTY BAG         2 2011-09-28 16:13:00   3.29   

        customer_id         country  
84016           NaN  United Kingdom  
97246           NaN  United Kingdom  
112723          NaN  United Kingdom  
116891          NaN  United Kingdom  
128107          NaN  United Kingdom  
160487          NaN  United Kingdom  
178669          NaN  United Kingdom  
262771          NaN  United Kingdom  
278378          NaN  United Kingdom  
365966          NaN  United Kingdom

invoice  stockcode      description  quantity         invoicedate  \
84017   543358  DCGSSGIRL  GIRLS PARTY BAG         3 2011-02-07 14:04:00   
112724  545897  DCGSSGIRL  GIRLS PARTY BAG         5 2011-03-07 17:22:00   
116892  546306  DCGSSGIRL  GIRLS PARTY BAG         2 2011-03-10 16:16:00   
128108  547249  DCGSSGIRL  GIRLS PARTY BAG        17 2011-03-22 09:27:00   
128269  547250  DCGSSGIRL  GIRLS PARTY BAG         2 2011-03-22 09:30:00   
150864  549524  DCGSSGIRL  GIRLS PARTY BAG         1 2011-04-08 15:42:00   
176169  551997  DCGSSGIRL  GIRLS PARTY BAG         2 2011-05-05 15:47:00   
178670  552234  DCGSSGIRL  GIRLS PARTY BAG         2 2011-05-06 16:01:00   
278379  561209  DCGSSGIRL  GIRLS PARTY BAG         5 2011-07-25 16:57:00   
297098  562933  DCGSSGIRL  GIRLS PARTY BAG         5 2011-08-10 16:51:00   

        price  customer_id         country  
84017    3.29          NaN  United Kingdom  
112724   3.29          NaN  United Kingdom  
116892   3.29          NaN  United Kingdom  
128108   3.29          NaN  United Kingdom  
128269   3.29          NaN  United Kingdom  
150864   3.29          NaN  United Kingdom  
176169   3.29          NaN  United Kingdom  
178670   3.29          NaN  United Kingdom  
278379   1.25          NaN  United Kingdom  
297098   3.29          NaN  United Kingdom

invoice     stockcode                         description  quantity  \
112442  545895  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
163439  550542  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
191910  553387  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
192071  553389  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
245515  558614  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
317441  564760  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   
317500  564762  gift_0001_10                                 NaN        30   
320440  564974  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         2   
536047  581217  gift_0001_10  Dotcomgiftshop Gift Voucher £10.00         1   

               invoicedate  price  customer_id         country  
112442 2011-03-07 17:14:00   8.33          NaN  United Kingdom  
163439 2011-04-19 11:37:00   8.33          NaN  United Kingdom  
191910 2011-05-16 16:33:00   8.33          NaN  United Kingdom  
192071 2011-05-16 16:37:00   8.33          NaN  United Kingdom  
245515 2011-06-30 15:56:00   8.33          NaN  United Kingdom  
317441 2011-08-30 10:47:00   8.33          NaN  United Kingdom  
317500 2011-08-30 10:48:00   0.00          NaN  United Kingdom  
320440 2011-08-31 15:32:00   8.33          NaN  United Kingdom  
536047 2011-12-08 09:20:00   8.33          NaN  United Kingdom

invoice stockcode                 description  quantity  \
157195  550193      PADS  PADS TO MATCH ALL CUSHIONS         1   
279045  561226      PADS  PADS TO MATCH ALL CUSHIONS         1   
358670  568158      PADS  PADS TO MATCH ALL CUSHIONS         1   
359871  568200      PADS  PADS TO MATCH ALL CUSHIONS         1   

               invoicedate  price  customer_id         country  
157195 2011-04-15 09:27:00  0.001      13952.0  United Kingdom  
279045 2011-07-26 10:13:00  0.001      15618.0  United Kingdom  
358670 2011-09-25 12:22:00  0.000      16133.0  United Kingdom  
359871 2011-09-25 14:58:00  0.001      16198.0  United Kingdom

invoice stockcode                 description  quantity  \
170783  551340  DCGS0004  HAYNES CAMPER SHOULDER BAG         1   

               invoicedate  price  customer_id         country  
170783 2011-04-27 17:23:00  16.63          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
279251  561249  DCGS0073        ebay        -4 2011-07-26 11:51:00    0.0   

        customer_id         country  
279251          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
279252  561250  DCGS0071         NaN        -2 2011-07-26 11:51:00    0.0   

        customer_id         country  
279252          NaN  United Kingdom

invoice  stockcode description  quantity         invoicedate  price  \
279254  561255  DCGS0066P         NaN        -3 2011-07-26 11:52:00    0.0   

        customer_id         country  
279254          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
279256  561253  DCGS0068        ebay       -10 2011-07-26 11:52:00    0.0   

        customer_id         country  
279256          NaN  United Kingdom

invoice stockcode description  quantity         invoicedate  price  \
279257  561254  DCGS0067        ebay       -11 2011-07-26 11:52:00    0.0   

        customer_id         country  
279257          NaN  United Kingdom

invoice stockcode      description  quantity         invoicedate  \
299982  A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   
299983  A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           price  customer_id         country  
299982  11062.06          NaN  United Kingdom  
299983 -11062.06          NaN  United Kingdom  
299984 -11062.06          NaN  United Kingdom

invoice stockcode      description  quantity         invoicedate  \
317508  C564763      CRUK  CRUK Commission        -1 2011-08-30 10:49:00   
324003  C565382      CRUK  CRUK Commission        -1 2011-09-02 15:45:00   
333745  C566216      CRUK  CRUK Commission        -1 2011-09-09 15:17:00   
338812  C566565      CRUK  CRUK Commission        -1 2011-09-13 12:32:00   
350582  C567655      CRUK  CRUK Commission        -1 2011-09-21 14:40:00   
361589  C568345      CRUK  CRUK Commission        -1 2011-09-26 15:28:00   
372785  C569245      CRUK  CRUK Commission        -1 2011-10-03 09:57:00   
389964  C570487      CRUK  CRUK Commission        -1 2011-10-10 17:12:00   
401767  C571440      CRUK  CRUK Commission        -1 2011-10-17 13:31:00   
416112  C572551      CRUK  CRUK Commission        -1 2011-10-24 17:07:00   

         price  customer_id         country  
317508    1.60      14096.0  United Kingdom  
324003   13.01      14096.0  United Kingdom  
333745   15.96      14096.0  United Kingdom  
338812   52.24      14096.0  United Kingdom  
350582  608.66      14096.0  United Kingdom  
361589  447.56      14096.0  United Kingdom  
372785  361.59      14096.0  United Kingdom  
389964  411.92      14096.0  United Kingdom  
401767  495.98      14096.0  United Kingdom  
416112  425.14      14096.0  United Kingdom

**`stockcode` code** | **Description** | **Implementation**
----------|----------|----------
`POST` | Postage charge. Delivery fees? | Exlcude
`D` | Discount applied | Exclude
`C2` | Shipping fee | Exclude
`DOT` | Dotcom purchase | Exclude
`M` | Manual purchase entrie | Exclude
`BANK CHARGES` | Bank fees | Payment processing fees
`S` | Sample product | Non-sale item
`AMAZONFEE` | Amazon marketplace fee | Third-party platform charge
`DCGS0076` | Gift voucher code 76 | Digital gift card
`DCGS0003` | Gift voucher code 3 | Digital gift card
`gift_0001_40` | £40 gift card | Fixed value gift card
`DCGS0070` | Gift voucher code 70 | Digital gift card
`m` | Manual adjustment (lowercase) | Administrative correction
`gift_0001_50` | £50 gift card | Fixed value gift card
`gift_0001_30` | £30 gift card | Fixed value gift card
`gift_0001_20` | £20 gift card | Fixed value gift card
`DCGS0055` | Gift voucher code 55 | Digital gift card
`DCGS0072` | Gift voucher code 72 | Digital gift card
`DCGS0074` | Gift voucher code 74 | Digital gift card
`DCGS0069` | Gift voucher code 69 | Digital gift card
`DCGS0057` | Gift voucher code 57 | Digital gift card
`DCGSSBOY` | Boy's gift voucher | Themed gift card
`DCGSSGIRL` | Girl's gift voucher | Themed gift card
`gift_0001_10` | £10 gift card | Fixed value gift card
`PADS` | Padding/packaging materials | Non-sale item
`DCGS0004` | Gift voucher code 4 | Digital gift card
`DCGS0073` | Gift voucher code 73 | Digital gift card
`DCGS0071` | Gift voucher code 71 | Digital gift card
`DCGS0066P` | Gift voucher code 66P | Digital gift card variant
`DCGS0068` | Gift voucher code 68 | Digital gift card
`DCGS0067` | Gift voucher code 67 | Digital gift card
`B` | Bad debt write-off | Financial adjustment
`CRUK` | Cancer Research UK donation | Charitable contribution

In [61]:
print("Number of duplicated rows:", df.duplicated().sum())

Number of duplicated rows: 5268
